plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergConnectorTest.java

```diff
@@ -5789,6 +5789,194 @@ public void testDropTableDeleteData()
+        testPartitionHiddenColumnTransform("hour(part)", "timestamp '2017-05-01 10:12:34'", "part_hour=2017-05-01-10", "timestamp '2018-05-01 10:12:34'", "part_hour=2018-05-01-10");
+        testPartitionHiddenColumnTransform("bucket(part, 10)", "1", "part_bucket=6", "2", "part_bucket=2");
+        testPartitionHiddenColumnTransform("truncate(part, 3)", "'abcde'", "part_trunc=abc", "'vwxyz'", "part_trunc=vwx");
+    }
+
+    private void testPartitionHiddenColumnTransform(String partitioning, String firstInput, String firstPartition, String secondInput, String secondPartition)
+    {
+        try (TestTable table = newTrinoTable("test_transform_partition", "WITH (partitioning = ARRAY['" + partitioning + "']) AS SELECT 1 id, " + firstInput + " part")) {
+            assertUpdate("INSERT INTO " + table.getName() + " VALUES (2, " + secondInput + ")", 1);
+
+            assertThat(computeActual("SELECT \"$partition\" FROM " + table.getName()).getOnlyColumnAsSet())
+                    .containsExactlyInAnyOrder(firstPartition, secondPartition);
+
+            assertThat(query("SELECT id FROM " + table.getName() + " WHERE \"$partition\" = '" + firstPartition + "'"))
+                    .isFullyPushedDown()
+                    .matches("VALUES 1");
+
+            assertThat(query("SELECT id FROM " + table.getName() + " WHERE \"$partition\" = '" + secondPartition + "'"))
+                    .isFullyPushedDown()
+                    .matches("VALUES 2");
+        }
+    }
+
+    @Test
+    void testPartitionHiddenColumnRenameColumn()
+    {
+        try (TestTable table = newTrinoTable("test_rename_partition", "WITH (partitioning = ARRAY['part']) AS SELECT 1 id, 10 part")) {
+            assertThat(query("SELECT id, \"$partition\" FROM " + table.getName()))
+                    .matches("VALUES (1, VARCHAR 'part=10')");
+
+            assertUpdate("ALTER TABLE " + table.getName() + " RENAME COLUMN part TO renamed_part");
+            assertThat(query("SELECT id, \"$partition\" FROM " + table.getName()))
+                    .matches("VALUES (1, VARCHAR 'part=10')");
+
+            assertUpdate("INSERT INTO " + table.getName() + " VALUES (2, 20)", 1);
+            assertThat(query("SELECT id, \"$partition\" FROM " + table.getName()))
+                    .matches("VALUES (1, VARCHAR 'part=10'), (2, VARCHAR 'part=20')");
+        }
+    }
+
+    @Test
+    void testPartitionHiddenColumnChangePartition()
+    {
+        try (TestTable table = newTrinoTable("test_change_partition", "WITH (partitioning = ARRAY['y']) AS SELECT 1 x, 10 y")) {
+            assertThat(query("SELECT x, \"$partition\" FROM " + table.getName()))
+                    .matches("VALUES (1, VARCHAR 'y=10')");
+
+            assertUpdate("ALTER TABLE " + table.getName() + " SET PROPERTIES partitioning = ARRAY['x']");
+            assertThat(query("SELECT x, \"$partition\" FROM " + table.getName()))
```

Comment:
Could you add test how query with partition predicate would look like after such evolution?